File: 2021/day02_1.py
from typing import NamedTuple
RAW ="""forward 5
down 5
forward 8
up 3
down 8
forward 2"""


class Instruction(NamedTuple):
    direction: str
    size: int

    @classmethod
    def from_string(cls, string):
        direction, size = string.split()
        return cls(direction, int(size))


class Ride:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def move(self, instruction: Instruction) -> None:
        if instruction.direction == "down":
            self.y += instruction.size
        elif instruction.direction == "up":
            self.y -= instruction.size
        elif instruction.direction == "forward":
            self.x += instruction.size
        else:
            raise RuntimeError

instructions  = [Instruction.from_string(line) for line in RAW.split("\n")]
ride = Ride(0, 0)
for i in instructions:
    ride.move(i)

assert ride.x * ride.y == 150


if __name__ == "__main__":
    with open("input/day02.txt") as f:
        instructions  = [Instruction.from_string(line) for line in f]

    ride = Ride(0, 0)
    for i in instructions:
        ride.move(i)

    print(ride.x * ride.y)

File: 2021/day02_1.go
package main

import (
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
	"strings"
)

type Instruction struct {
	direction string
	size      int
}

type Ride struct {
	x int
	y int
}

func (r *Ride) move(i Instruction) {
	if i.direction == "down" {
		r.y = r.y + i.size	
	} else if i.direction == "up" {
		r.y = r.y - i.size	
	} else if i.direction == "forward" {
		r.x = r.x + i.size	
	} else {
		log.Fatal("Unknown direction")
	}
}

type Ride2 struct {
	x int
	y int
	aim int
}

func (r *Ride2) move(i Instruction) {
	if i.direction == "down" {
		r.aim = r.aim + i.size	
	} else if i.direction == "up" {
		r.aim = r.aim - i.size	
	} else if i.direction == "forward" {
		r.x = r.x + i.size	
		r.y = r.y + i.size * r.aim
	} else {
		log.Fatal("Unknown direction")
	}
}

func main() {
	file, err := os.OpenFile("input/day02.txt", os.O_RDWR, 0644)
	if err != nil {
		fmt.Println("could not open file")
	}
	defer file.Close()
	contentBytes, err := io.ReadAll(file)
	lines := strings.Split(string(contentBytes), "\n")

	var instructions []Instruction
	
	for _, line := range lines {
		if len(line) == 0 {
			continue
		}
		words := strings.Split(line, " ")
		direction := words[0]
		size, _ := strconv.Atoi(words[1])
		instructions = append(instructions, Instruction{direction: direction, size: size})
	}
	
	ride := Ride{0, 0}
	ride2 := Ride2{0, 0, 0}
	for _, i := range instructions {
		ride.move(i)
		ride2.move(i)
	}
	fmt.Println(ride.x, ride.y, ride.x * ride.y)
	fmt.Println(ride2.x, ride2.y, ride2.x * ride2.y)
}
